README.md
# clickhouse-cargo
clickhouse-cargo accumulates insert queries and commit them to [clickhouse](https://clickhouse.yandex/) in batch jobs with retries on failure.

[Clickhouse is designed for batch data insertion with significant intervals](https://clickhouse.tech/docs/en/introduction/performance/#performance-when-inserting-data).
When inserting data to clickhouse from distributed node clusters, [some sort of centralized queue mechanism is required](https://github.com/ClickHouse/ClickHouse/issues/1067#issuecomment-320471793) in order to prevent Clickhouse from hitting the performance limit.
That brings complexity to the service chain and more difficulty to maintenance.

Clickhouse-cargo provides an easy way to batch data insertion to Clickhouse. It accumulates insert queries to local file caches, and commit them to Clickhouse in batch. Clickhouse-cargo will also automatically restore uncommitted local caches. That efficiently prevents data lose from unexpected Clickhouse accidents.

clickhouse-cargo 适用于分布式的 NodeJS 服务向 Clickhouse 频繁插入数据的应用场景。这个模块将向 Clickhouse 数据库的高频写入改为低频的批量插入。

## How it works

 1. A `cargo` instance accepts insert requests submitted by the `push` method and keep inserts in-memory.
 2. The `cargo` instance periodically flushs in-memory inserts to a file cache, then rotates this file and commits rotations to the Clickhouse server.
 3. In case of a Clickhouse commit failure, the cargo will retry the submission in the next round of its routine till the submission is successful.
 4. In case of the NodeJS process crash. in-memory inserts will be flushed immediately into the file cache.

### Cluster mode support

When running in cluster mode (such as [PM2 cluster deployment](https://pm2.keymetrics.io/docs/usage/cluster-mode/) ), all cargo workers will run through an election via udp communication @ 127.0.0.1:17888 to elect a leader worker. Then only the leader worker will carry on with file rotations and commitments.

## 工作原理

 1. `cargo` 实例接受 `push`方法所提交的插入请求，并将请求临时存放于内存中。
 1. `cargo` 周期性地将内存中累积的插入记录写入对应的文件缓存。随后将文件缓存进行滚动，并将滚出结果提交到 Clickhouse 数据库。
 4. 当向 Clickhouse 写入失败时，`cargo` 将会在下一轮检查周期中重试提交直到提交成功。
 5. 当本地的 NodeJS 进程奔溃时，内存中累积的插入请求会被同步写入对应的文件缓存。

### 支持集群模式

在集群模式下，所有的 cargo woker 将通过UDP通讯选举出一个领头的worker。 接着由这个领头的worker来负责文件缓存的滚动和提交到 Clickhouse 数据库。


## Install
```
$ npm install clickhouse-cargo
```

## Usage

```javascript
/*
sample table schema
`CREATE TABLE IF NOT EXISTS cargo_test.table_test
(
  \`time\` DateTime ,
  \`step\`  UInt32,
  \`pos_id\` String DEFAULT ''
)
ENGINE = Memory()`;
*/

const clickhouse-cargo = require("clickhouse-cargo");
const TABLE_NAME = `cargo_test.table_test`;
const NUM_OF_INSERTIONS = 27891; // NOTE: bulk flushs every 100 lines

// init clickhouse-cargo
clickhouse-cargo.init({
  "host":"play-api.clickhouse.tech",
  "port":"8443",
  "user":"playground",
  "password":"clickhouse"
});

// insert data
const theCargo = clickhouse-cargo.createCargo(TABLE_NAME);
for (let i =0, i < NUM_OF_INSERTIONS, i++){
  theCargo.push(new Date(), i, "string");
}
```

## API

### Initialization

__Init by code__


```javascript
clickhouse-cargo.init(options: Options)
```

*Options*

|                  | required | default       | description
| :--------------- | :------: | :------------ | :----------
| `host`           | ✓        |               | Host to connect.
| `cargoPath`      |          | `${cwd()}/cargo_files`              | Path to local cargo cache.
| `maxTime`        |          |  1000         | For how long in milliseconds, a cargo will keep in-memory insert buffer before flushing it to file.
| `maxRows`        |          |  100          | For how many rows a cargo will keep in-memory.
| `commitInterval` |          |  5000         | Interval(ms) for cargo to commit to ClickHouse.
| `maxInsetParts`  |          |  100          | For how many parts will be inserted into ClickHouse in a single exame routine. Keep value less then 300 to avoide `Too many parts` issue happend on clickhouse server-side
| `saveWhenCrash`  |          |  true         | When `false`, cargos will not flushSync in-memory data when node process crashes.
| `user`           |          |               | Authentication user.
| `password`       |          |               | Authentication password.
| `port`           |          | `8123`        | Server port number.
| `protocol`       |          | `'http:'`     | `'https:'` or `'http:'`.


__Init by the environment variable__

Init by the environment variable is recommended for real-world production.
Clickhouse-cargo recognises `process.env.CLICKHOUSE_CARGO_PROFILE` and seeks the config json file from `~/.clickhouse-cargo/${process.env.CLICKHOUSE_CARGO_PROFILE}`


### Create a Cargo instanse

```javascript
/*
* @param tableName String, the name of ClickHouse table which data is inserted
*/
const cargo = clickhouse-cargo.createCargo(tableName);
```

### Insert a row
```javascript
/*
Instead of inserting to Clickhouse directly, push row to the cargo. And the cargo will commit accumulated insertions to Clickhouse in batch.
*/
cargo.push(column0, column1, column2, column3...)
```






CHANGELOG.md

2021/6/24 Version: 3.1.1
===
 1. Reduce unnecessary log.

2021/6/24 Version: 3.1.0
===
 1. Cargo: now submits multiple(up to config.maxInsetParts) local uncommits in one http post request. This change solves "Too many parts" error on clickhouse server side.
    Note: use multistream to avoid memory leak when hundred of readable stream piping to the upload request.

2021/4/12 Version: 3.0.5
===
 1. Cargo: no longer extends EventEmitter

2021/4/12 Version: 3.0.3
===
 1. Introduce config.maxInsetParts a) to avoid local commits blocked due to a single failure in the commit queue. and b) to avoide `Too many parts` errors happend on clickhouse server-side

2021/3/15 Version: 3.0.2
===
 1. When multiple cargos are running on the same node, there is a large chance that only one cargo will be committed in each exame routine. Remove `Cargo::_isCommiting` lock since no need for that in the current sequential implementqation.
 2. Add handling to pm2 stop event. Refer: https://pm2.keymetrics.io/docs/usage/signals-clean-restart/







lib/test/03-commit-bulk.js
// Generated by CoffeeScript 2.5.1
(function() {
  var ClickHouse, INIT_OPTION, NUM_OF_LINE, STATEMENT_CREATE_TABLE, STATEMENT_DROP_TABLE, STATEMENT_SELECT, TABLE_NAME, assert, columnValueString, createCargo, debuglog, fs, getClickHouseClient, isInited, os, path, sleep;

  ({createCargo, isInited, getClickHouseClient} = require("../"));

  debuglog = require("debug")("chcargo:test:03");

  ({sleep} = require("../utils"));

  assert = require("assert");

  fs = require("fs");

  os = require("os");

  path = require("path");

  ClickHouse = require('@apla/clickhouse');

  TABLE_NAME = "cargo_test.unittest03";

  STATEMENT_CREATE_TABLE = `CREATE TABLE IF NOT EXISTS ${TABLE_NAME}
(
  \`time\` DateTime ,
  \`step\`  UInt32,
  \`pos_id\` String DEFAULT ''
)
ENGINE = Memory()`;

  STATEMENT_CREATE_TABLE = STATEMENT_CREATE_TABLE.replace(/\n|\r/g, ' ');

  STATEMENT_DROP_TABLE = `DROP TABLE IF EXISTS ${TABLE_NAME}`;

  columnValueString = Date.now().toString(36);

  STATEMENT_SELECT = `SELECT * FROM ${TABLE_NAME} WHERE pos_id='${columnValueString}' LIMIT 100000 FORMAT JSONCompactEachRow `;

  // refer
  INIT_OPTION = {
    host: "localhost",
    maxTime: 2000,
    maxRows: 100,
    commitInterval: 8000
  };

  NUM_OF_LINE = 27891; // NOTE: bulk flushs every 100 lines

  //NUM_OF_LINE = 9  # NOTE: bulk flushs every 100 lines
  getClickHouseClient = function() {
    var err, pathToConfig, profileConfig, profileName;
    profileName = process.env.CLICKHOUSE_CARGO_PROFILE;
    assert(profileName, "missing process.env.CLICKHOUSE_CARGO_PROFIL");
    if (path.extname(profileName) !== ".json") {
      profileName += ".json";
    }
    pathToConfig = path.join(os.homedir(), ".clickhouse-cargo", profileName);
    debuglog(`[getClickHouseClient] try auto init from CLICKHOUSE_CARGO_PROFILE from ${pathToConfig}`);
    try {
      profileConfig = JSON.parse(fs.readFileSync(pathToConfig));
    } catch (error) {
      err = error;
      debuglog("[static init] FAILED error:", err);
    }
    return new ClickHouse(profileConfig);
  };

  describe("commit bulk", function() {
    var clickHouseClient, theCargo;
    this.timeout(30000);
    clickHouseClient = getClickHouseClient();
    theCargo = null;
    before(function(done) {
      debuglog("[before]");
      clickHouseClient.query(STATEMENT_DROP_TABLE, function(err) {
        if (err != null) {
          throw err;
        }
        clickHouseClient.query(STATEMENT_CREATE_TABLE, function(err) {
          if (err != null) {
            throw err;
          }
          theCargo = createCargo(TABLE_NAME);
          if (fs.existsSync(theCargo.pathToCargoFile)) { // clean up existing log
            // receive notification when commit failed
            //theCargo.on 'error', (err)->
            //debuglog "[on cargo error] error:", err
            //return
            fs.unlinkSync(theCargo.pathToCargoFile);
          }
          done();
        });
      });
    });
    //after -> process.exit(0)
    it("push to cargo", async function() {
      var i, j, ref;
      for (i = j = 0, ref = NUM_OF_LINE; (0 <= ref ? j < ref : j > ref); i = 0 <= ref ? ++j : --j) {
        theCargo.push(new Date(), i, columnValueString);
      }
      await sleep(20); // wait file stream flush
    });
    it("bulk should committed", function(done) {
      assert(!fs.existsSync(theCargo.pathToCargoFile), "local file must be cleared");
      done();
    });
    //it "push to cargo", ->
    //for i in [0...NUM_OF_LINE]
    //theCargo.push(new Date, i, columnValueString)

    //await sleep 10 # wait file stream flush
    //return
    it("records should be written into remote ClickHouse server", function(done) {
      var rows, stream;
      rows = [];
      debuglog("[read db] select:", STATEMENT_SELECT);
      stream = clickHouseClient.query(STATEMENT_SELECT, {
        format: "JSONCompactEachRow"
      }, function(err, result) {
        var i, j, len, row;
        if (err != null) {
          done(err);
          return;
        }
        result = result.replace(/\n/g, ",").trim().replace(/,$/, '');
        result = `[ ${result} ]`;
        console.dir(result);
        result = JSON.parse(result);
        debuglog("[read db] result:", result.length);
        console.dir(result, {
          depth: 10
        });
        assert(result.length === NUM_OF_LINE, `unmatching row length local:${NUM_OF_LINE}, remote:${result.length}`);
        result.sort(function(a, b) {
          return a[1] - b[1];
        });
        for (i = j = 0, len = result.length; j < len; i = ++j) {
          row = result[i];
          assert(row[1] === i, "unmatching field 1 ");
          assert(row[2] === columnValueString, "unmatching field 2 ");
        }
        done();
      });
    });
  });

}).call(this);


lib/test/01-create-cargo.js
// Generated by CoffeeScript 2.5.1
(function() {
  var INIT_OPTION, QUERY, QUERY1, assert, createCargo, init, isInited;

  ({createCargo, isInited, init} = require("../"));

  assert = require("assert");

  INIT_OPTION = {
    host: "localhost",
    maxTime: 2000,
    maxRows: 100,
    commitInterval: 8000
  };

  QUERY = "test.cargo0";

  QUERY1 = "test.cargo1";

  describe("init clickhouse_cargo", function() {
    var cargo0;
    cargo0 = null;
    //before (done)->
    //init(INIT_OPTION)
    //done()
    //return

    //after -> process.exit(0)
    it("auto init when env set", function(done) {
      assert(isInited(), "should auto init when env set");
      done();
    });
    it("cargo must be created with tableName", function(done) {
      assert.throws((function() {
        return createCargo();
      }), Error, /tableName/);
      assert.throws((function() {
        return createCargo("select * from dual");
      }), Error, /tableName/);
      cargo0 = createCargo(QUERY);
      //console.log cargo0
      assert(cargo0);
      assert(cargo0.tableName === QUERY, "bad cargo0.tableName");
      //assert cargo0.maxTime is INIT_OPTION.maxTime, "bad cargo0.maxTime:#{cargo0.maxTime} => #{INIT_OPTION.maxTime}"
      //assert cargo0.maxRows is INIT_OPTION.maxRows, "bad cargo0.maxRows"
      //assert cargo0.commitInterval is INIT_OPTION.commitInterval, "bad cargo0.commitInterval"
      done();
    });
    it("create multiple cargoes with the same statement should result in one shared cargo", function(done) {
      var cargo1, cargo2, cargo3, cargo4;
      cargo1 = createCargo(QUERY);
      cargo2 = createCargo(QUERY);
      cargo3 = createCargo(QUERY);
      assert(((cargo0 === cargo1 && cargo1 === cargo2) && cargo2 === cargo3));
      cargo4 = createCargo(QUERY1);
      assert(cargo4 !== cargo3);
      done();
    });
  });

}).call(this);


lib/test/02-push-to-cargo.js
// Generated by CoffeeScript 2.5.1
(function() {
  var ClickHouse, INIT_OPTION, NUM_OF_LINE, STATEMENT_CREATE_TABLE, STATEMENT_DROP_TABLE, STATEMENT_INSERT, TABLE_NAME, _, assert, createCargo, debuglog, fs, isInited, os, path, sleep;

  ({createCargo, isInited} = require("../"));

  debuglog = require("debug")("chcargo:test:02");

  ({createCargo, isInited} = require("../"));

  assert = require("assert");

  fs = require("fs");

  os = require("os");

  path = require("path");

  _ = require("lodash");

  ClickHouse = require('@apla/clickhouse');

  ({sleep} = require("../utils"));

  TABLE_NAME = "cargo_test.unittest02";

  STATEMENT_INSERT = `INSERT INTO ${TABLE_NAME}`;

  STATEMENT_DROP_TABLE = `DROP TABLE IF EXISTS ${TABLE_NAME}`;

  STATEMENT_CREATE_TABLE = `CREATE TABLE IF NOT EXISTS ${TABLE_NAME}
(
  \`time\` DateTime ,
  \`step\`  UInt32,
  \`pos_id\` String DEFAULT ''
)
ENGINE = Memory()`;

  STATEMENT_CREATE_TABLE = STATEMENT_CREATE_TABLE.replace(/\n|\r/g, ' ');

  // refer
  INIT_OPTION = {
    host: "localhost",
    maxTime: 2000,
    maxRows: 100,
    commitInterval: 20000
  };

  NUM_OF_LINE = 429; // NOTE: bulk flushs every 100 lines

  describe("push log to cargo and flush manully", function() {
    var columnValueString, theCargo;
    this.timeout(20000);
    theCargo = createCargo(TABLE_NAME);
    if (fs.existsSync(theCargo.pathToCargoFile)) { // clean up existing log
      fs.unlinkSync(theCargo.pathToCargoFile);
    }
    columnValueString = Date.now().toString(36);
    it("push to cargo and flush manully", async function() {
      var i, j, ref;
      for (i = j = 0, ref = NUM_OF_LINE; (0 <= ref ? j < ref : j > ref); i = 0 <= ref ? ++j : --j) {
        theCargo.push(new Date(), i, columnValueString);
      }
      await theCargo.flushToFile();
    });
    it("cargo should flush to file", function() {
      assert(fs.existsSync(theCargo.pathToCargoFile), `log file not exist on ${theCargo.pathToCargoFile}`);
    });
    return it("exam content written on hd file", function(done) {
      var contentInHD, contentInHDArr, i, j, len, line;
      contentInHD = fs.readFileSync(theCargo.pathToCargoFile, 'utf8');
      //debuglog "[exam hd content] contentInHD:", contentInHD
      contentInHDArr = _.compact(contentInHD.split(/\r|\n|\r\n/));
      //debuglog "[exam hd content] contentInHDArr:", contentInHDArr
      assert(contentInHDArr.length === NUM_OF_LINE, `unmatching output length. NUM_OF_LINE:${NUM_OF_LINE}, contentInHDArr.length:${contentInHDArr.length}`);
      for (i = j = 0, len = contentInHDArr.length; j < len; i = ++j) {
        line = contentInHDArr[i];
        line = JSON.parse(line);
        //console.log line
        assert(line[1] === i, "unmatching field 1 ");
        assert(line[2] === columnValueString, "unmatching field 2 ");
      }
      done();
    });
  });

  describe("push log to cargo and flush automatically", function() {
    var columnValueString, theCargo;
    this.timeout(20000);
    theCargo = createCargo(`${TABLE_NAME}_set2`);
    if (fs.existsSync(theCargo.pathToCargoFile)) { // clean up existing log
      fs.unlinkSync(theCargo.pathToCargoFile);
    }
    columnValueString = Date.now().toString(36);
    it("push to cargo and flush manully", async function() {
      var i, j, ref;
      for (i = j = 0, ref = NUM_OF_LINE; (0 <= ref ? j < ref : j > ref); i = 0 <= ref ? ++j : --j) {
        theCargo.push(new Date(), i, columnValueString);
      }
      await sleep(5);
    });
    it("cargo should flush to file", function() {
      assert(fs.existsSync(theCargo.pathToCargoFile), `log file not exist on ${theCargo.pathToCargoFile}`);
    });
    return it("exam content written on hd file", function(done) {
      var contentInHD, contentInHDArr, i, j, len, line;
      contentInHD = fs.readFileSync(theCargo.pathToCargoFile, 'utf8');
      //debuglog "[exam hd content] contentInHD:", contentInHD
      contentInHDArr = _.compact(contentInHD.split(/\r|\n|\r\n/));
      debuglog("[exam hd content] contentInHDArr:", contentInHDArr);
      assert(contentInHDArr.length === NUM_OF_LINE, `unmatching output length. NUM_OF_LINE:${NUM_OF_LINE}, contentInHDArr.length:${contentInHDArr.length}`);
      for (i = j = 0, len = contentInHDArr.length; j < len; i = ++j) {
        line = contentInHDArr[i];
        line = JSON.parse(line);
        //console.log line
        assert(line[1] === i, "unmatching field 1 ");
        assert(line[2] === columnValueString, "unmatching field 2 ");
      }
      done();
    });
  });

}).call(this);


lib/bulk.js
// Generated by CoffeeScript 2.5.1
(function() {
  var Bulk, FILENAME_PREFIX, NOOP, NUM_OF_LINES_TO_CORK, StaticCountWithProcess, assert, cluster, debuglog, fs, path, toSQLDateString;

  cluster = require('cluster');

  path = require("path");

  fs = require("fs");

  assert = require("assert");

  debuglog = require("debug")("chcargo:bulk");

  StaticCountWithProcess = 0;

  // cork stream write
  NUM_OF_LINES_TO_CORK = 100;

  FILENAME_PREFIX = "bulk-";

  NOOP = function() {};

  toSQLDateString = function(date) {
    //debuglog "[toSQLDateString] date:", date
    return date.getUTCFullYear() + '-' + ('00' + (date.getUTCMonth() + 1)).slice(-2) + '-' + ('00' + date.getUTCDate()).slice(-2) + ' ' + ('00' + date.getUTCHours()).slice(-2) + ':' + ('00' + date.getUTCMinutes()).slice(-2) + ':' + ('00' + date.getUTCSeconds()).slice(-2);
  };

  Bulk = class Bulk {
    //toString : -> "[Bulk #{@id}@#{@pathToFile}]"
    toString() {
      return `[Bulk ${this.id}]`;
    }

    constructor(workingPath, presetId) {
      if (presetId) {
        //debuglog "[constructor] Bulk with presetId: #{presetId}"
        this.id = presetId;
        this.count = 1; // mark bulk has content when presetId given
      } else {
        this.id = Date.now().toString(36) + `_${++StaticCountWithProcess}`;
        if (cluster.isWorker) {
          // when launch as a worker by pm2
          this.id += `_${cluster.worker.id}`;
        }
        this.count = 0;
      }
      this.pathToFile = path.join(workingPath, FILENAME_PREFIX + this.id);
      debuglog(`[constructor] Bulk id: ${this.id}, cluster.worker.id:${cluster.worker && cluster.worker.id}, @pathToFile:`, this.pathToFile);
      this.outputStream = fs.createWriteStream(this.pathToFile, {
        flags: 'a'
      });
      // make sure writableStream is working
      this.outputStream.write("");
      this._committed = false;
      this._committing = false;
      return;
    }

    push(arr) {
      var i, item, j, len, line;
      if (!(Array.isArray(arr) && (arr.length > 0))) {
        debuglog(`${this} [push] empty arr`);
        return;
      }
      for (i = j = 0, len = arr.length; j < len; i = ++j) {
        item = arr[i];
        if (item instanceof Date) {
          arr[i] = toSQLDateString(item);
        }
      }
      line = JSON.stringify(arr);
      if (this.count % NUM_OF_LINES_TO_CORK === 0) {
        //debuglog "#{@} [push] line:", line

        // the primary intent of writable.cork() is to accommodate a situation in which several small chunks are written to the stream in rapid succession.
        this.outputStream.cork();
      }
      this.outputStream.write((this.count > 0 ? "\n" : "") + line, 'utf8');
      ++this.count;
      if (this.count % NUM_OF_LINES_TO_CORK === 0) {
        process.nextTick(() => {
          return this.outputStream.uncork();
        });
      }
    }

    // set the expiration of this bulk
    expire(ttl) {
      //debuglog "#{@} [expire] ttl:#{ttl}"
      ttl = parseInt(ttl) || 0;
      if (ttl < 1000) {
        ttl = 1000;
      }
      this.expireAt = Date.now() + ttl;
    }

    commit(clichouseClient, statement) {
      var theOutputStream;
      if (this._committing) {
        debuglog(`${this} [commit] IGNORE is _committing`);
        return;
      }
      assert(statement, "missing insert statment");
      // FIXME: quick fix for cluster mode conflicats
      if (!fs.existsSync(this.pathToFile)) {
        debuglog(`${this} [commit] FIXME cache file already removed.`, this.pathToFile);
        this._committed = true;
        this.outputStream.destroy();
        return;
      }
      this._committing = true; //lock
      debuglog(`${this} [commit] go committing, count:${this.count}, pathToFile:${this.pathToFile}`);
      theOutputStream = this.outputStream;
      theOutputStream.end((err) => {
        var bulkDataStream, dbStream;
        if (err != null) {
          debuglog(`${this} [commit] FAIL to end stream. error:`, err);
          this._committing = false; //unlock
          return;
        }
        bulkDataStream = fs.createReadStream(this.pathToFile);
        dbStream = clichouseClient.query(statement, {
          format: 'JSONCompactEachRow'
        });
        dbStream.on('error', (err) => {
          debuglog(`${this} [commit] FAILED error:`, err);
          this._committing = false;
        });
        dbStream.on('finish', () => {
          debuglog(`${this} [commit] success dbStream:finish`);
          this._committing = false;
          this._committed = true;
          fs.unlink(this.pathToFile, NOOP); // remove the physical file
        });
        bulkDataStream.pipe(dbStream);
      });
    }

    isExpired() {
      return (parseInt(this.expireAt) || 0) <= Date.now();
    }

    isEmpty() {
      return this.count === 0;
    }

    isCommitted() {
      return this._committed === true;
    }

  };

  Bulk.FILENAME_PREFIX = FILENAME_PREFIX;

  Bulk.toSQLDateString = toSQLDateString;

  module.exports = Bulk;

}).call(this);


lib/utils.js
// Generated by CoffeeScript 2.5.1
(function() {
  var assert, cargoOptionToHttpOption, http, https, pkg, sleep, toSQLDateString;

  pkg = require("../package.json");

  assert = require("assert");

  http = require('http');

  https = require('https');

  toSQLDateString = function(date) {
    //debuglog "[toSQLDateString] date:", date
    return date.getUTCFullYear() + '-' + ('00' + (date.getUTCMonth() + 1)).slice(-2) + '-' + ('00' + date.getUTCDate()).slice(-2) + ' ' + ('00' + date.getUTCHours()).slice(-2) + ':' + ('00' + date.getUTCMinutes()).slice(-2) + ':' + ('00' + date.getUTCSeconds()).slice(-2);
  };

  cargoOptionToHttpOption = function(cargoOption, mixin) {
    var options;
    options = {
      //protocol : if cargoOption.vehicle is https then 'https:' else 'http:'
      host: cargoOption.host,
      port: cargoOption.port,
      headers: {
        'X-ClickHouse-User': cargoOption.user,
        'X-ClickHouse-Key': cargoOption.password || '',
        'User-Agent': `${pkg.name}/${pkg.version}`
      }
    };
    return Object.assign(mixin || {}, options);
  };

  sleep = function(seconds) {
    console.log(`sleep ${seconds} seconds`);
    assert(Number.isInteger(seconds) && seconds > 0);
    return new Promise((resolve) => {
      return setTimeout(resolve, seconds * 1000);
    });
  };

  module.exports = {
    toSQLDateString: toSQLDateString,
    cargoOptionToHttpOption: cargoOptionToHttpOption,
    sleep: sleep
  };

}).call(this);


lib/leader_election.js
// Generated by CoffeeScript 2.5.1
(function() {
  // when in cluster worker mode, elect a leader for each cargo to avoid race condition when restoring existing bulks
  var CLUSTER_WORKER_ID, ClientSentCount, KnownHighestWorkerId, MAX_UDP_CONFIRM, REG_MSG_VALIDATOR, RemoteClientPortByCargoId, RemoteClientToPort, SERVICE_HOST, SERVICE_PORT, SERVICE_TYPE, UDPEchoSrv, assert, cluster, debuglog, dgram, err, getLeaderId, isThisLeader, msg, procSend, udpClient;

  assert = require("assert");

  cluster = require('cluster');

  dgram = require('dgram');

  CLUSTER_WORKER_ID = cluster.isMaster ? "nocluster" : cluster.worker.id;

  debuglog = require("debug")(`chcargo:leader_election@${CLUSTER_WORKER_ID}`);

  SERVICE_TYPE = "clickhouse-cargo";

  SERVICE_PORT = 17888;

  SERVICE_HOST = '127.0.0.1';

  REG_MSG_VALIDATOR = /^\d+@\w+/;

  MAX_UDP_CONFIRM = 10;

  RemoteClientPortByCargoId = {};

  RemoteClientToPort = {};

  if (cluster.isMaster) {
    isThisLeader = function() {
      return true;
    };
    getLeaderId = function() {
      return 0;
    };
  } else {
    // when running as a cluster
    KnownHighestWorkerId = cluster.worker.id;
    // setup upd server
    UDPEchoSrv = dgram.createSocket({
      type: 'udp4'
    });
    UDPEchoSrv.on('message', function(msg, rinfo) {
      var cargoId, port, workerId;
      msg = msg.toString('utf8');
      if (!REG_MSG_VALIDATOR.test(msg)) {
        debuglog(`[UDPEchoSrv] bad msg:${msg}`);
        return;
      }
      [workerId, cargoId] = msg.toString('utf8').split("@");
      ({port} = rinfo);
      //debuglog "[UDPEchoSrv > on msg] workerId:#{workerId}, cargoId:#{cargoId}, port:#{port}"
      RemoteClientToPort[workerId] = port;
//debuglog "[UDPEchoSrv > on msg] cargoId:#{cargoId}, RemoteClientToPort:", RemoteClientToPort
// echo back
      for (workerId in RemoteClientToPort) {
        port = RemoteClientToPort[workerId];
        port = parseInt(port);
        if (port !== SERVICE_PORT) {
          UDPEchoSrv.send(msg, 0, msg.length, port, SERVICE_HOST);
        }
      }
    });
    //debuglog "[UDPEchoSrv] echo back to worker:#{workerId}@port:#{port}"
    UDPEchoSrv.on("error", function(err) {
      debuglog("[static] UDP_ERR UDPEchoSrv error:", err);
    });
    try {
      UDPEchoSrv.bind(SERVICE_PORT, SERVICE_HOST, function() {
        debuglog("[static] UDPEchoSrv bind SUCCESS");
        UDPEchoSrv.setBroadcast(true);
      });
    } catch (error) {
      err = error;
      debuglog(`[static@${cluster.worker.id}] UDPEchoSrv bind failed. error`, err);
    }
    // setup udp client
    udpClient = dgram.createSocket("udp4");
    udpClient.on("message", function(msg) {
      var remoteCargoId, remoteWorkerId;
      msg = msg.toString('utf8');
      if (!REG_MSG_VALIDATOR.test(msg)) {
        debuglog(`[udpClient] bad msg:${msg}`);
        return;
      }
      [remoteWorkerId, remoteCargoId] = msg.toString('utf8').split("@");
      //debuglog "[udpClient] on msg remoteWorkerId:#{remoteWorkerId}, remoteCargoId:#{remoteCargoId}"
      remoteWorkerId = parseInt(remoteWorkerId) || 0;
      if (remoteWorkerId > KnownHighestWorkerId) {
        debuglog(`[udpClient] acknowledage new leader: ${KnownHighestWorkerId} -> ${remoteWorkerId}`);
        KnownHighestWorkerId = remoteWorkerId;
      }
    });
    ClientSentCount = 2;
    msg = Buffer.from(`${cluster.worker.id}@cluster_leader_election`);
    procSend = function() {
      if (!(ClientSentCount >= 64)) {
        ClientSentCount *= 2;
      }
      udpClient.send(msg, 0, msg.length, SERVICE_PORT, SERVICE_HOST);
      setTimeout(procSend, ClientSentCount);
    };
    procSend();
    isThisLeader = function() {
      return KnownHighestWorkerId === cluster.worker.id;
    };
    getLeaderId = function() {
      return KnownHighestWorkerId;
    };
  }

  module.exports = {
    isThisLeader: isThisLeader,
    getLeaderId: getLeaderId
  };

}).call(this);


lib/cluster-test/05-restore-bulks-when-cluster.js
// Generated by CoffeeScript 2.5.1
(function() {
  var ClickHouse, NUM_OF_LINE, STATEMENT_CREATE_TABLE, STATEMENT_DROP_TABLE, STATEMENT_INSERT2, StartCountWithinProcess, TABLE_NAME, assert, clickHouseClient, cluster, columnValueString, crashProc, createCargo, crypto, debuglog, fs, getClickHouseClient, isInited, os, path, prepareBulkCachFile, proc, theCargo;

  ({createCargo, isInited} = require("../"));

  debuglog = require("debug")("chcargo:cluster-test:05");

  assert = require("assert");

  crypto = require('crypto');

  fs = require("fs");

  os = require("os");

  path = require("path");

  cluster = require('cluster');

  ClickHouse = require('@apla/clickhouse');

  TABLE_NAME = "cargo_test.unittest05";

  STATEMENT_CREATE_TABLE = `CREATE TABLE IF NOT EXISTS ${TABLE_NAME}
(
  \`time\` DateTime ,
  \`step\`  UInt32,
  \`pos_id\` String DEFAULT ''
)
ENGINE = Memory()`;

  STATEMENT_CREATE_TABLE = STATEMENT_CREATE_TABLE.replace(/\n|\r/g, ' ');

  columnValueString = Date.now().toString(36);

  STATEMENT_DROP_TABLE = `DROP TABLE IF EXISTS ${TABLE_NAME}`;

  NUM_OF_LINE = 12;

  STATEMENT_INSERT2 = `insert into ${TABLE_NAME}`;

  getClickHouseClient = function() {
    var err, pathToConfig, profileConfig, profileName;
    profileName = process.env.CLICKHOUSE_CARGO_PROFILE;
    assert(profileName, "missing process.env.CLICKHOUSE_CARGO_PROFIL");
    if (path.extname(profileName) !== ".json") {
      profileName += ".json";
    }
    pathToConfig = path.join(os.homedir(), ".clickhouse-cargo", profileName);
    debuglog(`[getClickHouseClient] try auto init from CLICKHOUSE_CARGO_PROFILE from ${pathToConfig}`);
    try {
      profileConfig = JSON.parse(fs.readFileSync(pathToConfig));
    } catch (error) {
      err = error;
      debuglog("[static init] FAILED error:", err);
    }
    return new ClickHouse(profileConfig);
  };

  prepareBulkCachFile = function(tableName, label) {
    var arr, content, i, j, ref, theFilepath;
    theFilepath = path.join(process.cwd(), "cargo_files", `cargo_${tableName}.${Date.now().toString(36)}_unittest.nocluster.uncommitted`);
    debuglog("[prepare] theFilepath:", theFilepath);
    content = "";
    for (i = j = 0, ref = NUM_OF_LINE; (0 <= ref ? j < ref : j > ref); i = 0 <= ref ? ++j : --j) {
      arr = [Math.round(Date.now() / 1000), i, label || "cluster-cargo"];
      content += JSON.stringify(arr) + "\n";
    }
    content = content.substr(0, content.length - 1);
    fs.writeFileSync(theFilepath, content);
  };

  if (cluster.isMaster) {
    debuglog("[isMaster]");
    clickHouseClient = getClickHouseClient();
    clickHouseClient.query(STATEMENT_DROP_TABLE, function() {
      return clickHouseClient.query(STATEMENT_CREATE_TABLE, function() {
        var proc;
        prepareBulkCachFile(TABLE_NAME, "~batchAAA");
        prepareBulkCachFile(TABLE_NAME, "~batchBBB");
        proc = function() {
          var i, j;
// spawn worker
          for (i = j = 0; j < 8; i = ++j) {
            cluster.fork();
          }
        };
        proc();
        setInterval(proc, 20000);
      });
    });
  } else {
    //statment = not(cluster.worker.id % 2) && STATEMENT_INSERT || STATEMENT_INSERT2
    debuglog(`[isWorker ${cluster.worker.id}]`);
    //theCargo = createCargo(statment)
    theCargo = createCargo(TABLE_NAME);
    StartCountWithinProcess = 0;
    proc = function() {
      var i, j, mark, ref;
      mark = `worker@${cluster.worker.id}:${StartCountWithinProcess++}`;
      debuglog("proc insert: ", mark);
      for (i = j = 0, ref = NUM_OF_LINE; (0 <= ref ? j < ref : j > ref); i = 0 <= ref ? ++j : --j) {
        theCargo.push(new Date(), i, mark);
      }
    };
    setInterval(proc, 2000);
    crashProc = function() {
      nonExistingFunc();
    };
    setTimeout(crashProc, 10000 + Math.random() * 60000 >>> 0);
  }

}).call(this);
